### ISSUE TITLE
Is there a way to restrict specific variable names and specific function calls?

### ISSUE THREAD

**Issue by moroshko** [eslint/eslint#3497]
I'd like to check whether I set state in my (React) app. More specifically, I'd like to catch things like:
```js
this.state = { ... };
maybeSomethingElse.state = { ... };
this.setState({ ... });
maybeSomethingElse.setState({ ... });
```
In other words, I'd like to restrict setting `state` and calling `setState()`.

Can eslint help to catch these?

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by gyandeeps** [comment]
We have a plugin for react specific rules, have you tried it thier?
Link: https://github.com/yannickcr/eslint-plugin-react

**Comment by ilyavolodin** [comment]
To answer your question - yes, ESLint can help with that. However, that being said, I do not think this rule is generic enough to be added to the core. I suggest proposing it in `eslint-plugin-react` if it's a common pattern in React, or creating your own [plugin](http://eslint.org/docs/developer-guide/working-with-plugins) for a rule like that.

**Comment by moroshko** [comment]
@gyandeeps I couldn't find a rule in `eslint-plugin-react` that can help. But, probably worth raising an issue there.

@ilyavolodin I think it can be handy to have a core rule that doesn't allow specific function calls. For example, one might want to ensure that higher order methods like `.forEach()` and `.map()` are not used for performance reasons. You already have a specific rule for restricting `eval()` (`no-eval`). I think that a more generic rule will be really useful (it could default to `['eval']`).

**Comment by ilyavolodin** [comment]
@moroshko `no-eval` is there because evaluating code is a bad practice and can lead to performance and security issues.
We can't have a rule that absolutely requires configuration in order to do anything. Every rule that we have in ESLint can work with default values, but I can't think of any defaults for a rule like that. I think this just doesn't fit into core and should be created as a separate plugin.
If you want to proceed with plugin creation and run into issues, stop by our chat room, and we will help you create one.
